### ISSUE TITLE
[Meta Pseudo Labels] Hard or soft pseudo labels

### ISSUE THREAD

**Issue by the-klingspor** [google-research/google-research#584]
Thanks for the interesting paper,
I have a question concerning the pseudo labels that are computed using the teacher network. In section 2, subsection "Practical approximation" of v3 of your paper, you say that you use "hard pseudo labels from the teacher distribution" and that this prevents us form simply using backpropagation on equation (3). 
However, is it really the case that you use one-hot-encoded pseudo labels generated by the teacher to train the student? Looking at the code, it seems like you just compute the teacher output with "build_uda_crossentropy", store them in "logits" and use them as-is without a one-hot-encoding. Am I missing anything?
€dith: I think this is the relevant part of the code:
"
\# for backprop
    cross_entropy['s_on_u'] = tf.losses.softmax_cross_entropy(
        **onehot_labels=tf.stop_gradient(tf.nn.softmax(logits['u_aug'], -1)),**
        logits=logits['s_on_u'],
        label_smoothing=params.label_smoothing,
        reduction=tf.losses.Reduction.NONE)
"
Also, if we actually use soft pseudo labels, why do we still have to use the complicated gradient of gradient/Taylor approximation like in appendix A and can't just use the differentiability of equation (3) as implied in the paper? Maybe I'm just misunderstanding what you mean by "soft" and "hard" pseudo labels.

Thank you very much for your help

**Comment by the-klingspor** [comment]
I think I found the questions to most of my answers in #534 , sorry for the inconvenience.
